### PR TITLE
Add Python example with `MatrixCSR`

### PR DESCRIPTION
**Pull request by garth-wells**
Highly experimental, but shows sparse matrices which can have different types in a single program.

### CODE REVIEW

File: cpp/dolfinx/la/MatrixCSR.h [hunk 1]
```diff
@@ -224,40 +246,40 @@ class MatrixCSR
...
       // Columns indices for row
       std::int32_t row = rows[r];
-      if (row > max_row)
-        throw std::runtime_error("Local row out of range");
 
       // Current data row
       const T* xr = x.data() + r * cols.size();
 
+#ifndef NDEBUG
+      if (!(row + 1 < (int)_row_ptr.size()))
```
Comment (chrisrichardson) [inline cpp/dolfinx/la/MatrixCSR.h:273]: This is not quite right, I think. When calling set, it should not be allowed to set in the ghost rows, but _row_ptr contains the ghost rows too.

File: cpp/dolfinx/la/MatrixCSR.h [hunk 2]
```diff
@@ -266,32 +273,32 @@ class MatrixCSR
...
+      // Row index and current data row
       std::int32_t row = rows[r];
-      if (row > max_row)
-        throw std::runtime_error("Local row out of range");
-
-      // Current data row
       const T* xr = x.data() + r * cols.size();
 
+#ifndef NDEBUG
+      if (!(row + 1 < (int)_row_ptr.size()))
```
Comment (chrisrichardson) [inline cpp/dolfinx/la/MatrixCSR.h:298]:
```suggestion
      if (row >= (int)_row_ptr.size())
```

File: python/dolfinx/wrappers/fem.cpp [hunk 3]
```diff
@@ -138,6 +139,25 @@ void declare_functions(py::module& m)
...
       "and coefficients");
+  // Matrix
+  m.def(
+      "assemble_matrix",
+      [](dolfinx::la::MatrixCSR<T>& A, const dolfinx::fem::Form<T>& form,
+         const std::vector<std::shared_ptr<const dolfinx::fem::DirichletBC<T>>>&
+             bcs)
+      { dolfinx::fem::assemble_matrix<T>(A.mat_add_values(), form, bcs); },
+      "Experimental.");
+  // NOTE: This is relative slow. Need to investigate reasons
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/fem.cpp:150]: remove commented code?

File: python/dolfinx/wrappers/la.cpp [hunk 4]
```diff
@@ -156,6 +175,11 @@ void la(py::module& m)
       = m.def_submodule("petsc", "PETSc-specific linear algebra");
   petsc_module(petsc_mod);
 
+  // std::string pyclass_span_const_name = std::string("xltSpanConst_") +
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/la.cpp:178]: remove?

File: python/dolfinx/wrappers/la.cpp [hunk 5]
```diff
@@ -30,6 +34,10 @@ namespace
 template <typename T>
 void declare_objects(py::module& m, const std::string& type)
 {
+  // std::string pyclass_span_const_name = std::string("xltSpanConst_") + type;
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/la.cpp:37]: remove commented?

File: python/dolfinx/wrappers/la.cpp [hunk 6]
```diff
@@ -16,12 +16,16 @@
...
 #include <petsc4py/petsc4py.h>
 #include <pybind11/complex.h>
+#include <pybind11/functional.h>
 #include <pybind11/numpy.h>
 #include <pybind11/operators.h>
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
 #include <xtl/xspan.hpp>
 
+// PYBIND11_MAKE_OPAQUE(xtl::span<const std::int32_t>);
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/la.cpp:26]: remove commented

File: python/dolfinx/wrappers/la.cpp [hunk 7]
```diff
@@ -16,12 +16,16 @@
 #include <memory>
 #include <petsc4py/petsc4py.h>
 #include <pybind11/complex.h>
+#include <pybind11/functional.h>
```
Comment (chrisrichardson) [inline python/dolfinx/wrappers/la.cpp:19]: Is this needed?

**Comment by chrisrichardson** [comment]
Nice demonstration of mixing scalar types.

File: python/dolfinx/wrappers/fem.cpp (hunk 3 above)
Comment (garth-wells) [reply python/dolfinx/wrappers/fem.cpp:150]: I'd like to leave it in for the time being - might come back to it later.

**Pull request merged**